src/components/Dialog.tsx

```diff
@@ -14,57 +20,47 @@ type Props = {
   keyUp?: React.KeyboardEventHandler<HTMLDivElement>;
 };
 
-export default class Dialog extends Component<Props> {
-  titleRef: React.Ref<unknown>;
-  constructor(props: Props) {
-    super(props);
-    this.titleRef = React.createRef();
-  }
+const Dialog = (props: Props) => {
+  const { appElement, isOpen, closeFn, keyUp, body } = props;
+  var headerRef = useRef<HTMLHeadingElement>(null);
 
-  static propTypes = {
-    appElement: PropTypes.object.isRequired,
-    isOpen: PropTypes.bool.isRequired,
-    closeFn: PropTypes.func.isRequired,
-    keyUp: PropTypes.func,
-    body: PropTypes.oneOfType([PropTypes.bool, PropTypes.object]).isRequired,
-  };
+  useEffect(() => {
+    Modal.setAppElement(appElement);
+  }, []);
 
-  onKeyUp = (e: KeyboardEvent) => {
-    if (e.key === "Escape") this.props.closeFn();
+  const onKeyUp = (e: KeyboardEvent): void => {
+    if (e.key === "Escape") closeFn();
   };
 
-  componentDidMount() {
-    Modal.setAppElement(this.props.appElement);
-  }
+  // NOTE(Emmanuel): this feels like a level of indirection
+  // that refs are supposed to deal with.
+  const focusTitle = () => {
+    headerRef.current.focus();
+  };
 
-  focusTitle() {
-    document.getElementById("heading").focus();
-  }
```

Comment:
Using a ref is a great improvement over using an element id that is :crossed_fingers: hopefully unique!

Focusing is a surprisingly complicated thing in browsers. I found this article that gives a very thorough overview of all the ways you can focus things in react: https://blog.maisie.ink/react-ref-autofocus/

The one thing you might need to change here (depending on how `Modal` is implemented) is to wrap `focusTitle` in a `useCallback()` hook. i.e.:

```ts
const focusTitle = useCallback(() => headerRef.current.focus());
```

That will make it equivalent to the old version where `focusTitle` was a class method.

Reply:
The `useCallback` requirement is explained well here: https://blog.maisie.ink/react-ref-autofocus/#a-note-about-usecallback

Reply:
Thanks for the links - definitely helpful! I agree that useCallback is the safest option here, and I'm wondering if it should also be used in NodeEditable for the same reasons?

Reply:
I don't think it's necessary for NodeEditable because the callbacks only get called in response to user interaction and not on render. You could use it with NodeEditable, but it would only be a performance optimization.